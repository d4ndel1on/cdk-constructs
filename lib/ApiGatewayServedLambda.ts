import {Construct} from "constructs";
import {NodejsFunction, NodejsFunctionProps} from "aws-cdk-lib/aws-lambda-nodejs";
import {HttpApi} from "@aws-cdk/aws-apigatewayv2-alpha";
import * as path from "path";
import {BillingMode, Table, TableProps} from "aws-cdk-lib/aws-dynamodb";
import {CfnOutput, Duration} from "aws-cdk-lib";
import {Architecture, Runtime} from "aws-cdk-lib/aws-lambda";
import {RetentionDays} from "aws-cdk-lib/aws-logs";
import {HttpLambdaIntegration} from "@aws-cdk/aws-apigatewayv2-integrations-alpha";

export interface ApiGatewayServedLambdaProps {
  /**
   * Table properties to overwrite. No dynamodb table created on absent
   */
  table?: Pick<
    TableProps,
    'billingMode' |
    'partitionKey' |
    'sortKey' |
    'timeToLiveAttribute' |
    'removalPolicy' |
    'readCapacity' |
    'writeCapacity' |
    'tableName'
  >
  /**
   * Lambda properties to overwrite. Defaults are used if absent.
   */
  lambda?: Partial<Pick<
    NodejsFunctionProps,
    'runtime' |
    'timeout' |
    'memorySize' |
    'logRetention' |
    'handler' |
    'environment' |
    'entry'
  >>
}

export class ApiGatewayServedLambda extends Construct {

  readonly table: Table | undefined
  readonly lambda: NodejsFunction
  readonly api: HttpApi

  constructor(scope: Construct, id: string, props?: ApiGatewayServedLambdaProps) {
    super(scope, id);

    const lambda = props?.lambda
    const table = props?.table

    this.table = table ? new Table(this, 'Table', {
      billingMode: BillingMode.PAY_PER_REQUEST,
      ...table,
    }) : undefined

    this.lambda = new NodejsFunction(this, `${id}ApiLambda`, {
      runtime: Runtime.NODEJS_18_X,
      timeout: Duration.seconds(5),
      architecture: Architecture.ARM_64,
      memorySize: 128,
      entry: path.join(__dirname, '../src/api.ts'),
      logRetention: RetentionDays.ONE_WEEK,
      handler: 'handler',
      bundling: {
        minify: true,
        externalModules: ['@aws-sdk'],
      },
      ...lambda,
      environment: {
        TABLE: this.table?.tableName || '',
        ...lambda?.environment,
      },
    })

    this.table?.grantReadWriteData(this.lambda)

    this.api = new HttpApi(this, `${id}HttpApi`, {
      defaultIntegration: new HttpLambdaIntegration(`${id}ApiIntegration`, this.lambda),
    })

    new CfnOutput(this, `${id}ApiURL`, {
      value: this.api.apiEndpoint,
      exportName: `${id}:ApiURL`,
    })
    if (this.table) {
      new CfnOutput(this, `${id}DynamoDbTable`, {
        value: this.table.tableArn,
        exportName: `${id}:TableArn`,
      })
    }
  }
}