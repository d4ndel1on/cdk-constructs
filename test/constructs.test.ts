import {Stack} from "aws-cdk-lib";
import {ApiGatewayServedLambda} from "../lib/ApiGatewayServedLambda";
import {Template} from "aws-cdk-lib/assertions";
import {HostedReactApp} from "../lib/HostedReactApp";
import * as path from "path";
import {AttributeType} from "aws-cdk-lib/aws-dynamodb";

describe('ApiGatewayServedLambda', function () {
  it('should match snapshot', () => {
    const stack = new Stack()
    new ApiGatewayServedLambda(stack, 'Test', {
      lambda: {
        entry: path.join(__dirname, 'testHandler.ts')
      }
    })

    const template = Template.fromStack(stack)
    expect(template.toJSON()).toMatchSnapshot()
  })

  test('should match snapshot with dynamodb', () => {
    const stack = new Stack()
    new ApiGatewayServedLambda(stack, 'TestWithDynamodb', {
      lambda: {
        entry: path.join(__dirname, 'testHandler.ts')
      },
      table: {
        partitionKey: {type: AttributeType.STRING, name: 'key'},
        sortKey: {type: AttributeType.STRING, name: 'sort'},
      }
    })

    const template = Template.fromStack(stack)
    expect(template.toJSON()).toMatchSnapshot()
  })
})

describe('HostedReactApp', () => {
  it('should match snapshot', () => {
    const stack = new Stack()
    new HostedReactApp(stack, 'Test')

    const template = Template.fromStack(stack)
    expect(template.toJSON()).toMatchSnapshot()
  })
});
