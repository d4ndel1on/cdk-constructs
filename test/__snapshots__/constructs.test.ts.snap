// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ApiGatewayServedLambda should match snapshot 1`] = `
{
  "Mappings": {
    "DefaultCrNodeVersionMap": {
      "af-south-1": {
        "value": "nodejs16.x",
      },
      "ap-east-1": {
        "value": "nodejs16.x",
      },
      "ap-northeast-1": {
        "value": "nodejs16.x",
      },
      "ap-northeast-2": {
        "value": "nodejs16.x",
      },
      "ap-northeast-3": {
        "value": "nodejs16.x",
      },
      "ap-south-1": {
        "value": "nodejs16.x",
      },
      "ap-south-2": {
        "value": "nodejs16.x",
      },
      "ap-southeast-1": {
        "value": "nodejs16.x",
      },
      "ap-southeast-2": {
        "value": "nodejs16.x",
      },
      "ap-southeast-3": {
        "value": "nodejs16.x",
      },
      "ca-central-1": {
        "value": "nodejs16.x",
      },
      "cn-north-1": {
        "value": "nodejs16.x",
      },
      "cn-northwest-1": {
        "value": "nodejs16.x",
      },
      "eu-central-1": {
        "value": "nodejs16.x",
      },
      "eu-central-2": {
        "value": "nodejs16.x",
      },
      "eu-north-1": {
        "value": "nodejs16.x",
      },
      "eu-south-1": {
        "value": "nodejs16.x",
      },
      "eu-south-2": {
        "value": "nodejs16.x",
      },
      "eu-west-1": {
        "value": "nodejs16.x",
      },
      "eu-west-2": {
        "value": "nodejs16.x",
      },
      "eu-west-3": {
        "value": "nodejs16.x",
      },
      "me-central-1": {
        "value": "nodejs16.x",
      },
      "me-south-1": {
        "value": "nodejs16.x",
      },
      "sa-east-1": {
        "value": "nodejs16.x",
      },
      "us-east-1": {
        "value": "nodejs16.x",
      },
      "us-east-2": {
        "value": "nodejs16.x",
      },
      "us-gov-east-1": {
        "value": "nodejs16.x",
      },
      "us-gov-west-1": {
        "value": "nodejs16.x",
      },
      "us-iso-east-1": {
        "value": "nodejs14.x",
      },
      "us-iso-west-1": {
        "value": "nodejs14.x",
      },
      "us-isob-east-1": {
        "value": "nodejs14.x",
      },
      "us-west-1": {
        "value": "nodejs16.x",
      },
      "us-west-2": {
        "value": "nodejs16.x",
      },
    },
  },
  "Outputs": {
    "TestTestApiURLD106D81B": {
      "Export": {
        "Name": "Test:ApiURL",
      },
      "Value": {
        "Fn::GetAtt": [
          "TestTestHttpApi18F88D44",
          "ApiEndpoint",
        ],
      },
    },
  },
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aFD4BFC8A": {
      "DependsOn": [
        "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRoleDefaultPolicyADDA7DEB",
        "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "5fa1330271b8967d9254ba2d4a07144f8acefe8b77e6d6bba38261373a50d5f8.zip",
        },
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB",
            "Arn",
          ],
        },
        "Runtime": {
          "Fn::FindInMap": [
            "DefaultCrNodeVersionMap",
            {
              "Ref": "AWS::Region",
            },
            "value",
          ],
        },
      },
      "Type": "AWS::Lambda::Function",
    },
    "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRoleDefaultPolicyADDA7DEB": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "logs:PutRetentionPolicy",
                "logs:DeleteRetentionPolicy",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRoleDefaultPolicyADDA7DEB",
        "Roles": [
          {
            "Ref": "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "TestTestApiLambda61889FD0": {
      "DependsOn": [
        "TestTestApiLambdaServiceRole91D48CD9",
      ],
      "Properties": {
        "Architectures": [
          "arm64",
        ],
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "b36210762cc288990e845576fd86a290ba456007f5088a6863257207e851c1db.zip",
        },
        "Environment": {
          "Variables": {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
            "TABLE": "",
          },
        },
        "Handler": "index.handler",
        "MemorySize": 128,
        "Role": {
          "Fn::GetAtt": [
            "TestTestApiLambdaServiceRole91D48CD9",
            "Arn",
          ],
        },
        "Runtime": "nodejs18.x",
        "Timeout": 5,
      },
      "Type": "AWS::Lambda::Function",
    },
    "TestTestApiLambdaLogRetention45EF9C9C": {
      "Properties": {
        "LogGroupName": {
          "Fn::Join": [
            "",
            [
              "/aws/lambda/",
              {
                "Ref": "TestTestApiLambda61889FD0",
              },
            ],
          ],
        },
        "RetentionInDays": 7,
        "ServiceToken": {
          "Fn::GetAtt": [
            "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aFD4BFC8A",
            "Arn",
          ],
        },
      },
      "Type": "Custom::LogRetention",
    },
    "TestTestApiLambdaServiceRole91D48CD9": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "TestTestHttpApi18F88D44": {
      "Properties": {
        "Name": "TestHttpApi",
        "ProtocolType": "HTTP",
      },
      "Type": "AWS::ApiGatewayV2::Api",
    },
    "TestTestHttpApiDefaultRoute96498F77": {
      "Properties": {
        "ApiId": {
          "Ref": "TestTestHttpApi18F88D44",
        },
        "AuthorizationType": "NONE",
        "RouteKey": "$default",
        "Target": {
          "Fn::Join": [
            "",
            [
              "integrations/",
              {
                "Ref": "TestTestHttpApiDefaultRouteTestApiIntegrationE13BEF72",
              },
            ],
          ],
        },
      },
      "Type": "AWS::ApiGatewayV2::Route",
    },
    "TestTestHttpApiDefaultRouteTestApiIntegrationE13BEF72": {
      "Properties": {
        "ApiId": {
          "Ref": "TestTestHttpApi18F88D44",
        },
        "IntegrationType": "AWS_PROXY",
        "IntegrationUri": {
          "Fn::GetAtt": [
            "TestTestApiLambda61889FD0",
            "Arn",
          ],
        },
        "PayloadFormatVersion": "2.0",
      },
      "Type": "AWS::ApiGatewayV2::Integration",
    },
    "TestTestHttpApiDefaultRouteTestApiIntegrationPermission58157AB1": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "TestTestApiLambda61889FD0",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": {
          "Fn::Join": [
            "",
            [
              "arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              {
                "Ref": "AWS::Region",
              },
              ":",
              {
                "Ref": "AWS::AccountId",
              },
              ":",
              {
                "Ref": "TestTestHttpApi18F88D44",
              },
              "/*/*",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "TestTestHttpApiDefaultStage3E4AB17A": {
      "Properties": {
        "ApiId": {
          "Ref": "TestTestHttpApi18F88D44",
        },
        "AutoDeploy": true,
        "StageName": "$default",
      },
      "Type": "AWS::ApiGatewayV2::Stage",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;

exports[`ApiGatewayServedLambda should match snapshot with dynamodb 1`] = `
{
  "Mappings": {
    "DefaultCrNodeVersionMap": {
      "af-south-1": {
        "value": "nodejs16.x",
      },
      "ap-east-1": {
        "value": "nodejs16.x",
      },
      "ap-northeast-1": {
        "value": "nodejs16.x",
      },
      "ap-northeast-2": {
        "value": "nodejs16.x",
      },
      "ap-northeast-3": {
        "value": "nodejs16.x",
      },
      "ap-south-1": {
        "value": "nodejs16.x",
      },
      "ap-south-2": {
        "value": "nodejs16.x",
      },
      "ap-southeast-1": {
        "value": "nodejs16.x",
      },
      "ap-southeast-2": {
        "value": "nodejs16.x",
      },
      "ap-southeast-3": {
        "value": "nodejs16.x",
      },
      "ca-central-1": {
        "value": "nodejs16.x",
      },
      "cn-north-1": {
        "value": "nodejs16.x",
      },
      "cn-northwest-1": {
        "value": "nodejs16.x",
      },
      "eu-central-1": {
        "value": "nodejs16.x",
      },
      "eu-central-2": {
        "value": "nodejs16.x",
      },
      "eu-north-1": {
        "value": "nodejs16.x",
      },
      "eu-south-1": {
        "value": "nodejs16.x",
      },
      "eu-south-2": {
        "value": "nodejs16.x",
      },
      "eu-west-1": {
        "value": "nodejs16.x",
      },
      "eu-west-2": {
        "value": "nodejs16.x",
      },
      "eu-west-3": {
        "value": "nodejs16.x",
      },
      "me-central-1": {
        "value": "nodejs16.x",
      },
      "me-south-1": {
        "value": "nodejs16.x",
      },
      "sa-east-1": {
        "value": "nodejs16.x",
      },
      "us-east-1": {
        "value": "nodejs16.x",
      },
      "us-east-2": {
        "value": "nodejs16.x",
      },
      "us-gov-east-1": {
        "value": "nodejs16.x",
      },
      "us-gov-west-1": {
        "value": "nodejs16.x",
      },
      "us-iso-east-1": {
        "value": "nodejs14.x",
      },
      "us-iso-west-1": {
        "value": "nodejs14.x",
      },
      "us-isob-east-1": {
        "value": "nodejs14.x",
      },
      "us-west-1": {
        "value": "nodejs16.x",
      },
      "us-west-2": {
        "value": "nodejs16.x",
      },
    },
  },
  "Outputs": {
    "TestWithDynamodbTestWithDynamodbApiURL115421FD": {
      "Export": {
        "Name": "TestWithDynamodb:ApiURL",
      },
      "Value": {
        "Fn::GetAtt": [
          "TestWithDynamodbTestWithDynamodbHttpApi9F6270BE",
          "ApiEndpoint",
        ],
      },
    },
    "TestWithDynamodbTestWithDynamodbDynamoDbTable6BCB55D2": {
      "Export": {
        "Name": "TestWithDynamodb:TableArn",
      },
      "Value": {
        "Fn::GetAtt": [
          "TestWithDynamodbTable481B152A",
          "Arn",
        ],
      },
    },
  },
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aFD4BFC8A": {
      "DependsOn": [
        "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRoleDefaultPolicyADDA7DEB",
        "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "5fa1330271b8967d9254ba2d4a07144f8acefe8b77e6d6bba38261373a50d5f8.zip",
        },
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB",
            "Arn",
          ],
        },
        "Runtime": {
          "Fn::FindInMap": [
            "DefaultCrNodeVersionMap",
            {
              "Ref": "AWS::Region",
            },
            "value",
          ],
        },
      },
      "Type": "AWS::Lambda::Function",
    },
    "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRoleDefaultPolicyADDA7DEB": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "logs:PutRetentionPolicy",
                "logs:DeleteRetentionPolicy",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRoleDefaultPolicyADDA7DEB",
        "Roles": [
          {
            "Ref": "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "TestWithDynamodbTable481B152A": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "AttributeDefinitions": [
          {
            "AttributeName": "key",
            "AttributeType": "S",
          },
          {
            "AttributeName": "sort",
            "AttributeType": "S",
          },
        ],
        "BillingMode": "PAY_PER_REQUEST",
        "KeySchema": [
          {
            "AttributeName": "key",
            "KeyType": "HASH",
          },
          {
            "AttributeName": "sort",
            "KeyType": "RANGE",
          },
        ],
      },
      "Type": "AWS::DynamoDB::Table",
      "UpdateReplacePolicy": "Retain",
    },
    "TestWithDynamodbTestWithDynamodbApiLambdaFBB4AC0B": {
      "DependsOn": [
        "TestWithDynamodbTestWithDynamodbApiLambdaServiceRoleDefaultPolicyBB607C37",
        "TestWithDynamodbTestWithDynamodbApiLambdaServiceRole2F506E6D",
      ],
      "Properties": {
        "Architectures": [
          "arm64",
        ],
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "b36210762cc288990e845576fd86a290ba456007f5088a6863257207e851c1db.zip",
        },
        "Environment": {
          "Variables": {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
            "TABLE": {
              "Ref": "TestWithDynamodbTable481B152A",
            },
          },
        },
        "Handler": "index.handler",
        "MemorySize": 128,
        "Role": {
          "Fn::GetAtt": [
            "TestWithDynamodbTestWithDynamodbApiLambdaServiceRole2F506E6D",
            "Arn",
          ],
        },
        "Runtime": "nodejs18.x",
        "Timeout": 5,
      },
      "Type": "AWS::Lambda::Function",
    },
    "TestWithDynamodbTestWithDynamodbApiLambdaLogRetention16D075C6": {
      "Properties": {
        "LogGroupName": {
          "Fn::Join": [
            "",
            [
              "/aws/lambda/",
              {
                "Ref": "TestWithDynamodbTestWithDynamodbApiLambdaFBB4AC0B",
              },
            ],
          ],
        },
        "RetentionInDays": 7,
        "ServiceToken": {
          "Fn::GetAtt": [
            "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aFD4BFC8A",
            "Arn",
          ],
        },
      },
      "Type": "Custom::LogRetention",
    },
    "TestWithDynamodbTestWithDynamodbApiLambdaServiceRole2F506E6D": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "TestWithDynamodbTestWithDynamodbApiLambdaServiceRoleDefaultPolicyBB607C37": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "dynamodb:BatchGetItem",
                "dynamodb:GetRecords",
                "dynamodb:GetShardIterator",
                "dynamodb:Query",
                "dynamodb:GetItem",
                "dynamodb:Scan",
                "dynamodb:ConditionCheckItem",
                "dynamodb:BatchWriteItem",
                "dynamodb:PutItem",
                "dynamodb:UpdateItem",
                "dynamodb:DeleteItem",
                "dynamodb:DescribeTable",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "TestWithDynamodbTable481B152A",
                    "Arn",
                  ],
                },
                {
                  "Ref": "AWS::NoValue",
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "TestWithDynamodbTestWithDynamodbApiLambdaServiceRoleDefaultPolicyBB607C37",
        "Roles": [
          {
            "Ref": "TestWithDynamodbTestWithDynamodbApiLambdaServiceRole2F506E6D",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "TestWithDynamodbTestWithDynamodbHttpApi9F6270BE": {
      "Properties": {
        "Name": "TestWithDynamodbHttpApi",
        "ProtocolType": "HTTP",
      },
      "Type": "AWS::ApiGatewayV2::Api",
    },
    "TestWithDynamodbTestWithDynamodbHttpApiDefaultRouteBEADB2A9": {
      "Properties": {
        "ApiId": {
          "Ref": "TestWithDynamodbTestWithDynamodbHttpApi9F6270BE",
        },
        "AuthorizationType": "NONE",
        "RouteKey": "$default",
        "Target": {
          "Fn::Join": [
            "",
            [
              "integrations/",
              {
                "Ref": "TestWithDynamodbTestWithDynamodbHttpApiDefaultRouteTestWithDynamodbApiIntegration3D2ED7E4",
              },
            ],
          ],
        },
      },
      "Type": "AWS::ApiGatewayV2::Route",
    },
    "TestWithDynamodbTestWithDynamodbHttpApiDefaultRouteTestWithDynamodbApiIntegration3D2ED7E4": {
      "Properties": {
        "ApiId": {
          "Ref": "TestWithDynamodbTestWithDynamodbHttpApi9F6270BE",
        },
        "IntegrationType": "AWS_PROXY",
        "IntegrationUri": {
          "Fn::GetAtt": [
            "TestWithDynamodbTestWithDynamodbApiLambdaFBB4AC0B",
            "Arn",
          ],
        },
        "PayloadFormatVersion": "2.0",
      },
      "Type": "AWS::ApiGatewayV2::Integration",
    },
    "TestWithDynamodbTestWithDynamodbHttpApiDefaultRouteTestWithDynamodbApiIntegrationPermission5686FA86": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "TestWithDynamodbTestWithDynamodbApiLambdaFBB4AC0B",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": {
          "Fn::Join": [
            "",
            [
              "arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              {
                "Ref": "AWS::Region",
              },
              ":",
              {
                "Ref": "AWS::AccountId",
              },
              ":",
              {
                "Ref": "TestWithDynamodbTestWithDynamodbHttpApi9F6270BE",
              },
              "/*/*",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "TestWithDynamodbTestWithDynamodbHttpApiDefaultStage38A40BC2": {
      "Properties": {
        "ApiId": {
          "Ref": "TestWithDynamodbTestWithDynamodbHttpApi9F6270BE",
        },
        "AutoDeploy": true,
        "StageName": "$default",
      },
      "Type": "AWS::ApiGatewayV2::Stage",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;

exports[`HostedReactApp should match snapshot 1`] = `
{
  "Outputs": {
    "TestBucketName3D8D75EF": {
      "Description": "The name of the S3 bucket",
      "Export": {
        "Name": "Test:BucketName",
      },
      "Value": {
        "Ref": "TestTestFrontendBucket2AD9F23D",
      },
    },
    "TestCloudFrontURLB12EDD75": {
      "Description": "The distribution URL",
      "Export": {
        "Name": "Test:CloudfrontURL",
      },
      "Value": {
        "Fn::GetAtt": [
          "TestTestCloudfrontDistribution025BCC07",
          "DomainName",
        ],
      },
    },
  },
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "TestTestCloudfrontDistribution025BCC07": {
      "Properties": {
        "DistributionConfig": {
          "Comment": {
            "Fn::Join": [
              "",
              [
                "Test cloudfront distribution to serve ",
                {
                  "Ref": "TestTestFrontendBucket2AD9F23D",
                },
                " bucket",
              ],
            ],
          },
          "CustomErrorResponses": [
            {
              "ErrorCode": 404,
              "ResponseCode": 200,
              "ResponsePagePath": "/index.html",
            },
          ],
          "DefaultCacheBehavior": {
            "CachePolicyId": "658327ea-f89d-4fab-a63d-7e88639e58f6",
            "CachedMethods": [
              "GET",
              "HEAD",
            ],
            "Compress": true,
            "TargetOriginId": "TestTestCloudfrontDistributionOrigin155D00694",
            "ViewerProtocolPolicy": "redirect-to-https",
          },
          "DefaultRootObject": "index.html",
          "Enabled": true,
          "HttpVersion": "http2",
          "IPV6Enabled": true,
          "Origins": [
            {
              "DomainName": {
                "Fn::GetAtt": [
                  "TestTestFrontendBucket2AD9F23D",
                  "RegionalDomainName",
                ],
              },
              "Id": "TestTestCloudfrontDistributionOrigin155D00694",
              "S3OriginConfig": {
                "OriginAccessIdentity": {
                  "Fn::Join": [
                    "",
                    [
                      "origin-access-identity/cloudfront/",
                      {
                        "Ref": "TestTestCloudfrontDistributionOrigin1S3Origin1F2E47DE",
                      },
                    ],
                  ],
                },
              },
            },
          ],
          "PriceClass": "PriceClass_100",
        },
      },
      "Type": "AWS::CloudFront::Distribution",
    },
    "TestTestCloudfrontDistributionOrigin1S3Origin1F2E47DE": {
      "Properties": {
        "CloudFrontOriginAccessIdentityConfig": {
          "Comment": "Identity for TestTestCloudfrontDistributionOrigin155D00694",
        },
      },
      "Type": "AWS::CloudFront::CloudFrontOriginAccessIdentity",
    },
    "TestTestFrontendBucket2AD9F23D": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "PublicAccessBlockConfiguration": {
          "BlockPublicAcls": true,
          "BlockPublicPolicy": true,
          "IgnorePublicAcls": true,
          "RestrictPublicBuckets": true,
        },
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Delete",
    },
    "TestTestFrontendBucketPolicyFB2CF0AC": {
      "Properties": {
        "Bucket": {
          "Ref": "TestTestFrontendBucket2AD9F23D",
        },
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "s3:GetObject",
              "Effect": "Allow",
              "Principal": {
                "CanonicalUser": {
                  "Fn::GetAtt": [
                    "TestTestCloudfrontDistributionOrigin1S3Origin1F2E47DE",
                    "S3CanonicalUserId",
                  ],
                },
              },
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    {
                      "Fn::GetAtt": [
                        "TestTestFrontendBucket2AD9F23D",
                        "Arn",
                      ],
                    },
                    "/*",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::S3::BucketPolicy",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
